sound/soc/sof/ipc4-pcm.c

```diff
@@ -32,47 +32,123 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
+	pipeline = pipeline_widget->private;
+
+	/* do not add backend pipelines */
+	if (pipeline->is_backend)
+		goto src_add;
+
+	/* check if pipeline has already been added to the list */
+	for (i = 0 ; i < pipeline_info->count; i++)
+		if (pipeline_info->pipelines[i] == pipeline_widget)
+			goto src_add;
+
+	pipeline_count = pipeline_info->count++;
+	if (pipeline_info->count > SOF_IPC4_TRIGGER_MAX_PIPELINE_COUNT) {
+		dev_err(sdev->dev, "Cannot trigger %d pipelines at once\n", pipeline_info->count);
+		return -EINVAL;
+	}
+
+	pipeline_info->pipelines[pipeline_count] = pipeline_widget;
+
+src_add:
+	/* add pipelines in the source paths */
+	snd_soc_dapm_widget_for_each_source_path(widget, p) {
+		if (!widget_in_list(list, p->source))
+			continue;
+
+		if (!p->walking && p->source->dobj.private) {
+			p->walking = true;
+			ret = sof_ipc4_get_pipelines_in_order(sdev, p->source, list, pipeline_info);
+			p->walking = false;
+			if (ret < 0)
+				return ret;
+		}
+	}
+
+	return 0;
+}
+
 static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
 				      struct snd_pcm_substream *substream, int state)
 {
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
 	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct sof_ipc4_multi_pipeline_info pipeline_info;
 	struct snd_sof_widget *pipeline_widget;
 	struct snd_soc_dapm_widget_list *list;
 	struct snd_soc_dapm_widget *widget;
 	struct sof_ipc4_pipeline *pipeline;
 	struct snd_sof_widget *swidget;
 	struct snd_sof_pcm *spcm;
 	int ret = 0;
-	int num_widgets;
+	int num_widgets, i;
 
 	spcm = snd_sof_find_spcm_dai(component, rtd);
 	if (!spcm)
 		return -EINVAL;
 
 	list = spcm->stream[substream->stream].list;
 
+	pipeline_info.count = 0;
+
+	/* get the list of pipeline widgets that need to be triggered */
 	for_each_dapm_widgets(list, num_widgets, widget) {
 		swidget = widget->dobj.private;
 
 		if (!swidget)
 			continue;
 
-		/* find pipeline widget for the pipeline that this widget belongs to */
-		pipeline_widget = swidget->pipe_widget;
-		pipeline = pipeline_widget->private;
+		/*
+		 * IPC4 requires pipelines to be triggered in order starting at the sink and
+		 * walking all the way to the source. So the starting widget for playback is of
+		 * DAI type and for capture, it is of AIF type.
+		 */
+		if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK && !WIDGET_IS_DAI(widget->id))
+			continue;
 
-		/* skip triggering backend pipelines */
-		if (pipeline->is_backend || pipeline->state == state)
+		if (substream->stream == SNDRV_PCM_STREAM_CAPTURE && !WIDGET_IS_AIF(widget->id))
 			continue;
 
+		ret = sof_ipc4_get_pipelines_in_order(sdev, widget, list, &pipeline_info);
+		if (ret < 0)
+			return ret;
+	}
+
+	/*
+	 * trigger each pipeline separately and if any of the pipeline trigger fails, its state
+	 * will be reset in hw_free, so no need to rollback states here
+	 */
+	for (i = 0; i < pipeline_info.count; i++) {
```

Comment:
current sequence:
(1） pipeline 1 pause -> run, pipeline 2 pause -> run,  pipeline n pause -> run.
can we set multiple pipelines in one ipc msg like
(1) pipeline 1 ... n pause
(2) pipeline 1 ... n run

in current cSOF we deal with it like
````c
	if (state.extension.r.multi_ppl) {
		ppl_count = ppl_data->pipelines_count;
		ppl_id = ppl_data->ppl_id;
		dcache_invalidate_region((__sparse_force void __sparse_cache *)ppl_id,
					 sizeof(int) * ppl_count);
	} else {
		ppl_count = 1;
		id = state.primary.r.ppl_id;
		ppl_id = &id;
	}
````

Reply:
@RanderWang I used multiple pipeline IPC trigger and ran into all kinds of IPC errors with it. Thats why I updated this PR to remove the multiple IPC and replaced it with multiple IPCs

Reply:
actually I take that back @RanderWang. I messed up the pipeline data for multiple pipelines. Fixed now.